[1856] eliminate prime trains

philcampeau opened the pull request:
## Before clicking "Create"

- [x] Branch is derived from the latest `master`
- [x] Add the `pins` or `archive_alpha_games` label if this change will break existing games
- [x] Code passes linter with `docker compose exec rack rubocop -a`
- [x] Tests pass cleanly with `docker compose exec rack rake`

## Implementation Notes

### Explanation of Change

Long ago when 1856 was coded, rather than programming in a way for events to happen on specific trains other than the first of each rank, it was decided to create extra "prime" ranks for the final train of each type in order to trigger the float percentage change that happens in 1856 when the last train of a rank is purchased. 1836Jr56, which is built on 1856, also used this method.

I had to make a few changes for this to work. 

- eliminated the prime ranks
- added an additional train to each rank that previously had a prime rank
- added events to each rank triggering the change on the final train
- These new events cluttered the game info page, ~so I added a new HIDDEN_EVENTS constant to /game/base.rb and tweaked the game_info.rb code to exclude any events listed in HIDDEN_EVENTS.~
  - This is changed to a check in `View::Game::GameInfo` for a `hidden` key in the event hash definition. If this key is present and the value is truthy then the event will not be shown.
- I took the opportunity to clean up some code in the game_trains and game_phases methods.

~Unfortunately, every attempt I made at writing a migration script resulted in some drift of the loans. I'm not sure why, the scripts I wrote were only changing the train ids, but every time the corporations would end up with a different amount of money at some point which would result in a failure when they couldn't afford a train that they had purchased in the game.~

~I've done test runs through several games and haven't witnessed this in my own plays, so I suspect the issue is more to do with the way loans are handled in the 1856 code itself, which is pretty wonky if I'm being honest.~

~**All that to say, we'll need to pin all existing games of 1856 and 1836Jr56**, and I've had to delete the fixture files. Once some new games are completed on the site, we can use them to create new fixtures, of course.~

**🏴‍☠️ This pull request has been hijacked by @ollybh 🏴‍☠️**

A migration has been added that can migrate all games of 1836Jr56, and all 1856 games except one (that has a train bought from the bank pool whose identifier is changed). The fixture files have been updated to work with these changes.

This PR does not touch any 18KA code. That game is derived from 1856 so will also need updating if it is resurrected – it is still in pre-alpha status and hasn't had any significant development committed in a few years.

philcampeau committed:
[1856] eliminates need for prime trains

philcampeau committed:
corrects an unnecessary change I had left in

philcampeau committed:
code cleanup

michaeljb commented:
I don't like needing to pin all of the instances of a "production" game, and I don't want to remove all the fixtures to replace them later--in fact in #12083 I'm working on adding spec tests that require fixtures are present for all released games.

I was able to edit `51745.json` by hand to "migrate" it:
* `"train": "2'-0"` -> `"train": "2-5"` and `"variant": "2'"` -> `"variant": "2"`
* `"train": "3'-0"` -> `"train": "3-4"` and so on

So I think a migration script ought to be possible.

michaeljb reviewed (changes requested):
I don't like needing to pin all of the instances of a prod game, and I don't want to remove all the fixtures to replace them later--in fact in #12083 I'm working on adding spec tests that require fixtures are present for all released games.

I was able to edit `51745.json` by hand to "migrate" it:
* `"train": "2'-0"` -> `"train": "2-5"` and `"variant": "2'"` -> `"variant": "2"`
* `"train": "3'-0"` -> `"train": "3-4"` and so on

So I think a migration script ought to be possible.

<img width="1582" height="1034" alt="Screenshot 2025-10-02 at 21 36 34" src="https://github.com/user-attachments/assets/d3d3e745-0e18-4ef6-900d-1731dff96381" />
<img width="1582" height="1034" alt="Screenshot 2025-10-02 at 21 36 36" src="https://github.com/user-attachments/assets/0ce9e947-94cd-4010-b7c9-8d7101506eca" />

philcampeau commented:
The issue comes up after the merger, which happens on the 6-train. Looking at all the fixture files, not a single one goes to the 6-train. 

I'll keep tinkering and see what I can find.

ollybh committed:
[1856/1836Jr56] Restore fixtures and spec file

This partially reverts commit 77392d36cd0b7a2398222d0be7c00da3e06866fb.

ollybh committed:
Merge branch 'upstream' into 1856-eliminate-prime-trains

ollybh reviewed (changes requested):
Like @michaeljb said, this needs a migration script to avoid pinning almost every single game. This shouldn't be complicated and I tried putting one together.

The first issue I ran into is that this does not change to having full capitalisation once the first 6-train is available. Capitalisation is determined from the phase status array, and phase 5 has an `incremental` status. `event_no_more_incremental_corps!` doesn't do anything other than printing a log message.

ollybh committed:
[1836Jr56] Initialise `@percent_to_operate`

Commit 77392d3 changed the handling of the variable float percentages in
the 1856 code, introducing a `@percent_to_operate` variable to the game
class. 1836Jr56 depends on the 1856 code, but this variable was not
being initialised, causing failures.

ollybh committed:
[1856/1836Jr56] Fix corporation capitalisation

In 1856 the changes to corporation capitalisation does not match up with
the phase changes: the change from escrow to incremental capitalisation
is when the first five train becomes available for purchase, and the
change from incremental to full capitalisation is when the first six
train becomes available.

This has been implemented with the 4' and 5' phases but these have been
eliminated. Instead track the capitalisation state in a
`@capitalization` variable, which is updated in events tied to the
purchase of the last 4-train and 5-train.

ollybh committed:
[1856/1836Jr56] Fixed fixture files following removal of prime trains

The prime trains (2'-0, 3'-0, etc) no longer exist. They have been
replaced by normal trains (2-5, 3-4, etc). This amends the existing
fixture files to use the new train identifiers.

ollybh committed:
[1856/1836Jr56] Migration script for removal of prime trains

This is a script to edit the game database following the removal of the
prime trains. It replaces the train identifiers and train variants.

ollybh committed:
[Core/1856/1836Jr56] Hidden events that are not shown on game info page

An alternate implementation of hidden events. These are ones that are
not listed on the game info page.

Instead of a global `HIDDEN_EVENTS` array, this is done by adding a
`hidden` key to the event hash. If this is a truthy value then the event
will not be shown.

ollybh committed:
[1856] Make attribute read-only

`percent_to_operate` is only modified within the game class, so make it
a read-only attribute.

ollybh committed:
[Info page] Revert small refactoring

Reverting this change (from commit 77392d3) to reduce the code churn.
IMHO it didn't make the code any clearer.

philcampeau commented:
Damn dude! This is great. I was away on vacation the last couple of weeks. I'll take a loot at this tomorrow. Thanks!

ollybh committed:
[1856] Fix typo

ollybh reviewed (approved):
Getting rid of my request for changes flag on this PR.

crericha reviewed (approved):
@ollybh migration script looks good to me; however, if you have any concerns, maybe import a few games locally and try the script on them before the entire prod db?

michaeljb commented:
Running the migration locally worked for 361/363 games. Will pin those two and migrate the rest.

[222637](https://18xx.games/game/222637) failed during the migration:

```
Game 222637 [1856]: failed validation Not a buyable train: 4-0
```

[245040](https://18xx.games/game/245040) appeared to pass the migration but failed during subsequent validation:

```
#<Engine::GameError: Blocking step Buy Trains cannot process action pass at 411>
```
